pallets/circuit-breaker/src/lib.rs

```diff
@@ -0,0 +1,470 @@
+		/// Represented as a non-zero fraction (nominator, denominator) with the max value being 10_000.
+		type DefaultMaxNetTradeVolumeLimitPerBlock: Get<(u32, u32)>;
+
+		/// The maximum percentage of a pool's liquidity that can be added in a block.
+		/// Represented as an optional non-zero fraction (nominator, denominator) with the max value being 10_000.
+		/// If set to None, the limits are not enforced.
+		type DefaultMaxLiquidityLimitPerBlock: Get<Option<(u32, u32)>>;
+
+		/// Omnipool's hub asset id. The limits are not tracked for this asses.
+		type OmnipoolHubAsset: Get<Self::AssetId>;
+
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	/// Default maximum net trade volume limit per block
+	#[pallet::type_value]
+	pub fn DefaultTradeVolumeLimit<T: Config>() -> (u32, u32) {
+		T::DefaultMaxNetTradeVolumeLimitPerBlock::get()
+	}
+
+	#[pallet::storage]
+	/// Trade volume limits of assets that don't use the default value
+	#[pallet::getter(fn trade_volume_limit_per_asset)]
+	pub type TradeVolumeLimitPerAsset<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AssetId, (u32, u32), ValueQuery, DefaultTradeVolumeLimit<T>>;
+
+	#[pallet::storage]
+	/// Trade volumes per asset
+	#[pallet::getter(fn allowed_trade_volume_limit_per_asset)]
+	pub type AllowedTradeVolumeLimitPerAsset<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AssetId, TradeVolumeLimit<T>>;
+
+	/// Default maximum liquidity limit per block
+	#[pallet::type_value]
+	pub fn DefaultLiquidityLimit<T: Config>() -> Option<(u32, u32)> {
+		T::DefaultMaxLiquidityLimitPerBlock::get()
+	}
+
+	#[pallet::storage]
+	/// Liquidity limits of assets that don't use the default value
+	#[pallet::getter(fn liquidity_limit_per_asset)]
+	pub type LiquidityLimitPerAsset<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AssetId, Option<(u32, u32)>, ValueQuery, DefaultLiquidityLimit<T>>;
+
+	#[pallet::storage]
+	/// Liquidity volumes per asset
+	#[pallet::getter(fn allowed_liquidity_limit_per_asset)]
+	pub type AllowedLiquidityAmountPerAsset<T: Config> = StorageMap<_, Blake2_128Concat, T::AssetId, LiquidityLimit<T>>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Trade volume limit of an asset was changed.
+		TradeVolumeLimitChanged {
+			asset_id: T::AssetId,
+			trade_volume_limit: (u32, u32),
+		},
+		/// Liquidity limit of an asset was changed.
+		LiquidityLimitChanged {
+			asset_id: T::AssetId,
+			liquidity_limit: Option<(u32, u32)>,
+		},
+	}
+
+	#[pallet::error]
+	#[cfg_attr(test, derive(PartialEq, Eq))]
+	pub enum Error<T> {
+		/// Invalid value for a limit. Limit must be non-zero.
+		InvalidLimitValue,
+		/// Allowed liquidity limit is not stored for asset
+		LiquidityLimitNotStoredForAsset,
```

Comment:
In the previous implementation, there were two handlers. The first one for registering assets, the second one for updating the values. So it was possible call the second handler before registering the assets. In the current implementation we use just one handler, so it's not a problem anymore. But I decided to keep this error because the getter method returns `Option`.

Reply:
not sure about this. Limit is calculated from an asset reserve. We can use `0` or `1` as the default value, something what prevents users from doing operations on it.

Reply:
sry i confused this with a limit config (which has default value), still this error shouldn't happen because the pair should be always initialized already. therefore i would use something like genric `InconsistentState` for it rather than named pallet event